Add phpunit 9 support (to allow testing on PHP 8)

ADmad opened the pull request:
<!---

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.

The best way to propose a feature is to open an issue first and discuss your ideas there before implementing them.

Always follow the [contribution guidelines](https://github.com/cakephp/cakephp/blob/master/.github/CONTRIBUTING.md) when submitting a pull request. In particular, make sure existing tests still pass, and add tests for all new behavior. When fixing a bug, you may want to add a test to verify the fix.

-->

ADmad committed:
Add support for PHPUnit 9.3 (which support PHP 8)

othercorey commented:
```
3) Cake\Test\TestCase\Database\Type\DateTypeTest::testMarshal with data set #9 ('2014-02-14', Cake\Chronos\Date Object (...))
Failed asserting that two DateTime objects are equal.
--- Expected
+++ Actual
@@ @@
-2014-02-14T00:00:00.000000+0000
+2014-02-14T07:42:50.000000+0000
```

This looks like a fun failure.

ADmad commented:
Errors on PHP 8:

- > Function openssl_free_key() is deprecated

    I couldn't find any info regarding this. I tried replacing it with `openssl_pkey_free()` which is apparently it's alias but that function too gives deprecation error. So nothing sure if there's some alternative or we just need to drop it's usage for PHP 8.

- > Function libxml_disable_entity_loader() is deprecated

    Based on https://github.com/php/php-src/pull/5867 and https://github.com/symfony/symfony/pull/37763 loading external entities is disabled by default since v2.9. So the solution is to check libxml version and use the method only for < 2.9. For higher versions use `LIBXML_NOENT` flag instead of enable.

- PHP version constraint for https://github.com/auraphp/Aura.Intl/blob/3.x/composer.json#L23 would have to be bumped. Aura project seem to be dead. I'll make a PR and see how it goes.

ADmad commented:
Last commit to Aura.Intl was on Jan 20, 2017. Even running it's test suite with a recent phpunit version would be non trivial. I think it would be better if we just copy the classes we need and move on.

othercorey commented:
>   I couldn't find any info regarding this. I tried replacing it with `openssl_pkey_free()` which is apparently it's alias but that function too gives deprecation error. So nothing sure if there's some alternative or we just need to drop it's usage for PHP 8.

>> The openssl_pkey_free() function is deprecated and no longer has an effect,
 instead the OpenSSLAsymmetricKey instance is automatically destroyed if it is no
 longer referenced.

https://github.com/php/php-src/blob/master/UPGRADING#L397

ADmad commented:
> https://github.com/php/php-src/blob/master/UPGRADING#L397

Good, so just need to wrap the call in a PHP version check.

ADmad committed:
openssl_free_key() is deprecated and unneeded as of PHP 8

othercorey commented:
We can commit this with failures allowed so people can test fixes or will have to wait work on them here.

ADmad commented:
On #14885 is merged and forwarded to `4.next` the build on PHP 8 will start passing so there's no need to modify the travis matrix.

ADmad committed:
Add new assertion methods added in PHP 9.

This allows our testsuite to be used with PHP 8.5+ & 9+ without
deprecation warnings.

ADmad committed:
Avoid relying on wacky PHP behavior by comparing non numeric string casted to float.

The check behaves different in PHP 8 causing test failures.

garas commented on src/Validation/Validation.php:
```diff
@@ -315,19 +322,21 @@ public static function numElements($check, string $operator, int $expectedCount)
-     *    less or equal <=, is less <, equal to ==, not equal !=
+     * @param string $operator Can be one of following operator strings:
+     *   '>', '<', '>=', '<=', '==', '!=', '===' and '!=='. You can use one of
+     *   the Validation::COMPARISON_* constants.
```
```suggestion
     *   the Validation::COMPARE_* constants.
```

othercorey commented on .travis.yml:
```diff
@@ -64,7 +61,7 @@ install:
       if [[ $PREFER_LOWEST == 1 ]]; then
         composer update --prefer-lowest --prefer-stable --no-interaction
       elif [[ $TRAVIS_PHP_VERSION == 'nightly' ]]; then
-        composer install --no-interaction --ignore-platform-reqs
```
What does no interaction cause with php 8?

othercorey reviewed (approved):
Looks ok to me.

ADmad commented:
This is ready to merge now. 

The only remaining failure for DateTimeType marshalling needs a fix in Chronos (which isn't an easy fix) and dealing with the warnings for usage of `at()` for mocks will take significant amount of effort given the number of usages. Perhaps this would be a good time to start switching to Mockery instead.
